systemd/systemd.go

```diff
@@ -593,33 +597,82 @@ func (s *systemd) getGlobalUserStatus(unitNames ...string) ([]*UnitStatus, error
 	return sts, nil
 }
 
-func (s *systemd) CurrentMemoryUsage(unit string) (quantity.Size, error) {
-	out, err := s.systemctl("show", "--property", "MemoryCurrent", unit)
+func (s *systemd) getPropertyStringValue(unit, key string) (string, error) {
+	// XXX: ignore stderr of systemctl command to avoid further infractions
+	//      around LP #1885597
+	out, err := s.systemctl("show", "--property", key, unit)
 	if err != nil {
-		return 0, osutil.OutputErr(out, err)
+		return "", osutil.OutputErr(out, err)
 	}
-	// strip the MemoryCurrent= from the output
-	splitVal := strings.SplitN(strings.TrimSpace(string(out)), "=", 2)
+	cleanVal := strings.TrimSpace(string(out))
+
+	// strip the TasksCurrent= from the output
+	splitVal := strings.SplitN(cleanVal, "=", 2)
 	if len(splitVal) != 2 {
-		return 0, fmt.Errorf("invalid property format from systemd for MemoryCurrent")
+		return "", fmt.Errorf("invalid property format from systemd for %s (got %s)", key, cleanVal)
+	}
+
+	return strings.TrimSpace(splitVal[1]), nil
+}
+
+func (s *systemd) CurrentTasksCount(unit string) (int, error) {
+	tasksStr, err := s.getPropertyStringValue(unit, "TasksCurrent")
+	if err != nil {
+		return 0, err
 	}
 
-	trimmedVal := strings.TrimSpace(splitVal[1])
+	// if the unit is inactive or doesn't exist, the tasks current can be
+	// reported as "[not set]"
+	if tasksStr == "[not set]" {
+		return 0, fmt.Errorf("tasks count unavailable")
+	}
+
+	intVal, err := strconv.Atoi(tasksStr)
+	if err != nil {
+		return 0, fmt.Errorf("invalid property value from systemd for TasksCurrent: cannot parse %q as an integer", tasksStr)
+	}
+
+	return intVal, nil
+}
+
+func (s *systemd) CurrentMemoryUsage(unit string) (quantity.Size, error) {
+	memStr, err := s.getPropertyStringValue(unit, "MemoryCurrent")
```

Comment:
I played a bit with this on CentOS 7, so far the findings suggest that when MemoryLimit is not set (meaning there is no group under memory controller hierarchy), MemoryCurrent reported by systemd has an absurd value in exabytes. At least adding MemoryAccounting=true makes the problem go away, as the group is created and memory stats are there. So the bogus value is observed only when there's no kernel source for the statisics and seems to be some internal default value in systemd. I was able to reproduce the same exact behavior for TasksCurrent, that is setting TasksAccounting=true makes the reported value be sane, while restarting the slice without accounting, is again a very high default value.

I suspect that the systemd behavior may have changed and with unified hierarchy a process is always part of some slice which has memory & task accounting enabled (which afaik is always true with cgroup v2).

Bit erratic log from my experiments: https://paste.ubuntu.com/p/TfYRHQTqWn/

Reply:
it sound like we should set MemoryAccounting=true in our slices anyway?

Reply:
but we have always set MemoryAccounting=true in the slice units we generate ?